### ISSUE TITLE
feat(tensor_parallel): complete RowParallelLinear implementation

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#117]
## 问题描述
`deepspeed/core/tensor_parallel/layers.py` 中 `RowParallelLinear` 需完成以下 DES-LOC 扩展：
- 输入维度（input_size）非整除 TP 时的 dummy-row padding（pad input_size 到 tp 倍数）
- `input_is_parallel=False` 路径：scatter_to_tensor_parallel_region 前需处理 padding mask
- `sharded_state_dict()` 接口与 dist_checkpointing 对接（weight axis=1 shard）
- `is_expert=True` 路径下 expert-parallel RNG 初始化

## 技术方案
1. `__init__` 中：`padded_input_size = ceil(input_size / tp_world_size) * tp_world_size`，本 rank 的 input slice = `padded_input_size // tp_world_size`
2. `forward()` 中：若 `not input_is_parallel`，在 scatter 前对输入 pad 到 `padded_input_size`；每 rank forward 完毕做 reduce-scatter/all-reduce，最后 rank 截掉 dummy 部分（bias 不受影响）
3. 实现 `sharded_state_dict()`：weight 在 axis=1 打 TP shard tag
4. 连接 `autosp_hook.register_tp_hooks()` 确保 desloc_engine 可注册 RowParallelLinear
5. 测试文件：`tests/unit/core/tensor_parallel/test_row_parallel_linear.py`

## 验收条件
```bash
# 1. 导入检查
python -c "from deepspeed.core.tensor_parallel.layers import RowParallelLinear; print('OK')"

# 2. 无残留 stub
grep -n "NotImplementedError" deepspeed/core/tensor_parallel/layers.py | grep -i "row" && echo "FAIL" || echo "PASS"

# 3. heterogeneous padding 单测（input_size=512, tp=3）
python -m pytest tests/unit/core/tensor_parallel/test_row_parallel_linear.py -k "hetero" -v

# 4. input_is_parallel=False 路径覆盖
python -m pytest tests/unit/core/tensor_parallel/test_row_parallel_linear.py -k "scatter" -v

# 5. sharded_state_dict 接口
python -c "from deepspeed.core.tensor_parallel.layers import RowParallelLinear; assert 'sharded_state_dict' in dir(RowParallelLinear); print('OK')"
```

## 优先级
P1 — RowParallelLinear 与 ColumnParallelLinear 并列为 TP 核心，blocking #127 (optimizer wiring)
